services/bundle_analysis/new_notify/contexts/__init__.py

```diff
@@ -0,0 +1,152 @@
+from functools import cached_property
+
+from shared.bundle_analysis import (
+    BundleAnalysisReport,
+    BundleAnalysisReportLoader,
+)
+from shared.torngit.base import TorngitBaseAdapter
+from shared.yaml import UserYaml
+
+from database.enums import ReportType
+from database.models.core import Commit, Repository
+from database.models.reports import CommitReport
+from services.archive import ArchiveService
+from services.bundle_analysis.new_notify.types import NotificationType
+from services.repository import (
+    get_repo_provider_service,
+)
+from services.storage import get_storage_client
+
+
+class BaseBundleAnalysisNotificationContext:
+    """Base NotificationContext for bundle analysis notifications.
+    It includes basic information that all bundle analysis notifications need.
+    Use NotificationContextBuilder to populate the context.
+
+    Example:
+      builder = NotificationContextBuilder(commit, current_yaml, GITHUB_APP_INSTALLATION_DEFAULT_NAME)
+      notification_context = builder.build_context().get_result()
+    """
+
+    notification_type: NotificationType
+
+    def __init__(
+        self, commit: Commit, current_yaml: UserYaml, gh_app_installation_name: str
+    ) -> None:
+        self.commit = commit
+        self.current_yaml = current_yaml
+        self.gh_app_installation_name = gh_app_installation_name
+
+    @cached_property
+    def repository(self) -> Repository:
+        return self.commit.repository
+
+    @cached_property
+    def repository_service(self) -> TorngitBaseAdapter:
+        return get_repo_provider_service(
+            self.repository,
+            installation_name_to_use=self.gh_app_installation_name,
+        )
+
+    @property
+    def commit_report(self) -> CommitReport:
```

Comment:
For this and other properties that have a getter/setter, wouldn't accessing the property without setting a value for it first lead to an AttributeError? I think we should add some validation to each, something like 

``` python
@property
def commit_report(self) -> CommitReport:
    if self._commit_report is None:
        raise ValueError("...")
    return self._commit_report
```

Or an alternative would be to add None placeholders for properties that aren't initialized in the constructor, like 

```
    def __init__(
        self, commit: Commit, current_yaml: UserYaml, gh_app_installation_name: str
    ) -> None:
        self.commit = commit
        self.current_yaml = current_yaml
        self.gh_app_installation_name = gh_app_installation_name
        self._commit_report = None
```

Or, there's the approach of not using getters/setters if we choose the None route above. It would be a bit unclearer to know where that property gets populated, but reading this doc https://realpython.com/python-getter-setter/ gave me the impression that the main point of the setters here would be to add additional requirements/behaviors to the value being set, while we're just setting the property to a value here. 

I'm okay with whichever style we choose to get/set these properties outside of the constructor, but I would make sure we have a way to deal with accessing a non-initialized variable, like commit_report here

Reply:
> While this is currently used by the Builder class where it deals with this somewhat, nothing prevents this class being used elsewhere in the future so just doing some damage control for then 👌

I think the whole idea of having a [builder pattern](https://refactoring.guru/design-patterns/builder) is exactly to exclusively use the builder to create the object, and use it in a more `read_only` manner.
Which means that the creating of the Context shouldn't happen without the builder (except possibly for tests, but we are careful with those)

But yeah, you are correct that it's not clear if you happen to use the context without initializing it why this happened.
I will follow the 1st suggestion of having a custom error that indicates "you should build the Context using the builder before using it"